File: LeetCode/Study/ArrayAndString/16.SmallestLengthSubarray.py
"""
给定一个含有 n 个正整数的数组和一个正整数 target 。
找出该数组中满足其总和大于等于 target 的长度最小的 子数组 [numsl, numsl+1, ..., numsr-1, numsr] ，并返回其长度。如果不存在符合条件的子数组，返回 0 。
"""

from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        min_length = float('inf')  # 初始化最小长度为无穷大
        current_sum = 0  # 当前窗口的和
        left = 0  # 滑动窗口的左边界

        # 遍历数组，右边界逐步扩大
        for right in range(n):
            current_sum += nums[right]  # 扩大窗口，累加右边界的元素

            # 当窗口和大于等于目标值时，尝试缩小窗口
            while current_sum >= target:
                # 更新最小长度
                min_length = min(min_length, right - left + 1)
                current_sum -= nums[left]  # 缩小窗口，移动左边界
                left += 1  # 移动左边界，减少窗口和

        # 如果找到了满足条件的子数组，返回最小长度，否则返回 0
        return min_length if min_length != float('inf') else 0


def test():
    sol = Solution()
    test_cases = [
        (7, [2, 3, 1, 2, 4, 3], 2),  # 最小子数组长度是 2, 子数组是 [4, 3]
        (4, [1, 4, 4], 1),  # 最小子数组长度是 1, 子数组是 [4]
        (11, [1, 1, 1, 1, 1, 1, 1, 1], 0),  # 没有满足条件的子数组
        (15, [1, 2, 3, 4, 5], 5),  # 最小子数组长度是 5, 子数组是 [1, 2, 3, 4, 5]
        (8, [1, 1, 1, 1, 1, 1, 1, 1], 8),  # 最小子数组长度是 8, 子数组是整个数组
    ]
    
    for i, (target, nums, expected) in enumerate(test_cases, 1):
        result = sol.minSubArrayLen(target, nums)
        print(f"Test case {i}:")
        print(f"Target: {target}, Nums: {nums}")
        print(f"Expected: {expected}, Result: {result}")
        print("✅ Pass" if result == expected else "❌ Fail")
        print("-" * 40)

if __name__ == "__main__":
    test()

File: LeetCode/Study/ArrayAndString/16.SmallestLengthSubarray.java
/*
 * 给定一个含有 n 个正整数的数组和一个正整数 target 。
 * 找出该数组中满足其总和大于等于 target 的长度最小的 子数组 [numsl, numsl+1, ..., numsr-1, numsr] ，并返回其长度。
 * 如果不存在符合条件的子数组，返回 0 。
 */

import java.util.Arrays;

class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int minLength = Integer.MAX_VALUE, left = 0, current_sum = 0;

        for (int right = 0; right < nums.length; right++) {
            current_sum += nums[right]; // 扩大窗口，增加右边界

            // 当当前窗口的和大于等于目标值时，尝试缩小窗口
            while (current_sum >= target) {
                minLength = Math.min(minLength, right - left + 1); // 更新最小长度
                current_sum -= nums[left]; // 缩小窗口，减少左边界
                left++;
            }
        }

        return minLength == Integer.MAX_VALUE ? 0 : minLength; // 如果没有找到满足条件的子数组，返回0
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[][] testArrays = {
            {2, 3, 1, 2, 4, 3}, // expected: 2
            {1, 4, 4},          // expected: 1
            {1, 1, 1, 1, 1, 1, 1, 1}, // expected: 0
            {1, 2, 3, 4, 5},    // expected: 2
            {5, 1, 3, 5, 10, 7, 4, 9, 2, 8} // expected: 2
        };
        int[] targets = {
            7,
            4,
            11,
            11,
            15
        };
        int[] expected = {
            2,
            1,
            0,
            3,
            2
        };

        for (int i = 0; i < testArrays.length; i++) {
            int result = sol.minSubArrayLen(targets[i], testArrays[i]);
            System.out.println("Test " + (i + 1));
            System.out.println("Input array: " + Arrays.toString(testArrays[i]));
            System.out.println("Target: " + targets[i]);
            System.out.println("Expected: " + expected[i] + ", Got: " + result);
            System.out.println(result == expected[i] ? "✅ Pass" : "❌ Fail");
            System.out.println("-".repeat(40));
        }
    }
}

